### ISSUE TITLE
模型训练问题

### ISSUE THREAD

**Issue by Lee-inso** [ViTAE-Transformer/ViTAE-Transformer-Remote-Sensing#23]
用的mmseg原生的potsdam.py预处理的
参照#9 用的potsdam_ori.py
config
upernet_swin_tiny_patch4_window7_512x512_80k_potsdam
参数未改
batch_size = 8
model = dict(
               pretrained='checkpoint/upernet-rsp-swin-t-potsdam-latest.pth')
pipline跟data中的
reduce_zero_label = True

2023-03-10 00:13:21,947 - mmseg - INFO - Iter [80000/80000]     lr: 7.500e-10, eta: 0:00:00, time: 0.378, data_time: 0.003, memory: 15450, decode.loss_ce: 0.1454, decode.acc_seg: 74.9332, aux.loss_ce: 0.0674, aux.acc_seg: 74.0879, loss: 0.2129

+--------------------+-------+-------+--------+-----------+--------+
|       Class        |  IoU  |  Acc  | Fscore | Precision | Recall |
+--------------------+-------+-------+--------+-----------+--------+
| impervious_surface | 82.19 | 92.91 | 90.23  |   87.69   | 92.91  |
|      building      | 91.04 | 97.03 | 95.31  |   93.64   | 97.03  |
|   low_vegetation   | 70.99 |  88.3 | 83.03  |   78.36   |  88.3  |
|        tree        |  73.7 | 83.01 | 84.86  |    86.8   | 83.01  |
|        car         | 81.17 | 89.05 | 89.61  |   90.17   | 89.05  |
|      clutter       |  0.0  |  0.0  |  nan   |    nan    |  0.0   |
+--------------------+-------+-------+--------+-----------+--------+
2023-03-10 00:13:54,226 - mmseg - INFO - Summary:
2023-03-10 00:13:54,226 - mmseg - INFO - 
+-------+-------+-------+---------+------------+---------+
|  aAcc |  mIoU |  mAcc | mFscore | mPrecision | mRecall |
+-------+-------+-------+---------+------------+---------+
| 86.86 | 66.52 | 75.05 |  88.61  |   87.33    |  75.05  |
+-------+-------+-------+---------+------------+---------+

F1 88.61 
请问，哪里出问题了么
感谢:)

**Comment by DotWang** [comment]
@Lee-inso 
这个问题比较复杂，需要慢慢盘： 
1. 你用的是哪种数据和标签？
数据有，RGB, IRRG和RGBIR三种
标签有'5_Labels_all.zip'和'5_Labels_all_noBoundary.zip'两种
2. mmseg用'5_Labels_all_noBoundary.zip'，它首先将黑边转换到0，其余颜色转换到1-6，然后`reduce_zero_label=True`等一通操作后将黑色转到255，然后通过`ignore_index=255`在算loss时不考虑黑色，然后mmseg直接评测其余六个类别（包含clutter）的精度
3. 我们用'5_Labels_all.zip'，其中无黑色边界，因此我们直接将类别转到0-5（5是clutter），保持`reduce_zero_label=False`即可，因为遥感论文中通常不考虑cluster，所以我们`num_classes=5` `ignore_index=5`
4. 如果你用'5_Labels_all_noBoundary.zip'，用mmseg转完以后，会有包含黑色和clutter在内七个类别，这样无法同时ignore两个类别
5. 如果你用'5_Labels_all.zip'，用mmseg转完以后，因为无黑色，类别变成1-6，所以需要`reduce_zero_label=True`，变成0-5，然后你需要同时在config和potsdam_ori.py中(如果你用的话)设置`ignore_index=5`

所以我猜，虽然你用了potsdam_ori.py， 但是没有设置`ignore_index=5`，影响了精度的计算

**Comment by DotWang** [comment]
@Lee-inso 你这个预训练权重好像也不对吧，应该用仓库首页的readme，见下图，那个分割文件夹readme中的模型是训练好的分割网络

![image](https://user-images.githubusercontent.com/35402675/224225371-6e91c2f9-3da1-44ce-b5cc-8600bef71bab.png)

**Comment by Xie-Muxi** [comment]
不行诶
“
如果你用'5_Labels_all.zip'，用mmseg转完以后，因为无黑色，类别变成1-6，所以需要reduce_zero_label=True，变成0-5，然后你需要同时在config和potsdam_ori.py中(如果你用的话)设置ignore_index=5

”
我是用MMSEG自带的划分，因为我想做一些控制变量的基准测试，这是我修改的config

```python

norm_cfg = dict(type='SyncBN', requires_grad=True)
model = dict(
    type='EncoderDecoder',
    pretrained=
    '/root/PycharmProjects/xmx/ViTAE-Transformer-Remote-Sensing/SemanticSegmentation/model/rsp-vitaev2-s-ckpt.pth',
    backbone=dict(
        type='ViTAE_Window_NoShift_basic',
        RC_tokens_type=['swin', 'swin', 'transformer', 'transformer'],
        NC_tokens_type=['swin', 'swin', 'transformer', 'transformer'],
        stages=4,
        embed_dims=[64, 64, 128, 256],
        token_dims=[64, 128, 256, 512],
        downsample_ratios=[4, 2, 2, 2],
        NC_depth=[2, 2, 8, 2],
        NC_heads=[1, 2, 4, 8],
        RC_heads=[1, 1, 2, 4],
        mlp_ratio=4.0,
        NC_group=[1, 32, 64, 128],
        RC_group=[1, 16, 32, 64],
        img_size=512,
        window_size=7,
        drop_path_rate=0.3),
    decode_head=dict(
        type='UPerHead',
        in_channels=[64, 128, 256, 512],
        in_index=[0, 1, 2, 3],
        pool_scales=(1, 2, 3, 6),
        channels=512,
        dropout_ratio=0.1,
        num_classes=5,
        norm_cfg=dict(type='SyncBN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
        ignore_index=5),
    auxiliary_head=dict(
        type='FCNHead',
        in_channels=256,
        in_index=2,
        channels=256,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=5,
        norm_cfg=dict(type='SyncBN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4),
        ignore_index=5),
    train_cfg=dict(),
    test_cfg=dict(mode='whole'))
dataset_type = 'PotsdamDataset'
data_root = '/workspace/CV/users/wangdi153/Dataset/potsdam_dataset/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
crop_size = (512, 512)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', reduce_zero_label=True),
    dict(type='Resize', img_scale=(512, 512), ratio_range=(0.5, 2.0)),
    dict(type='RandomCrop', crop_size=(512, 512), cat_max_ratio=0.75),
    dict(type='RandomFlip', prob=0.5),
    dict(type='PhotoMetricDistortion'),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='Pad', size=(512, 512), pad_val=0, seg_pad_val=5),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg'])
]
val_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(512, 512),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=None,
        img_ratios=[1.0],
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=4,
    train=dict(
        type='PotsdamDataset',
        data_root='/root/PycharmProjects/xmx/data/potsdam_all/',
        img_dir='img_dir_IRRG/train',
        ann_dir='ann_dir_IRRG/train',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations', reduce_zero_label=True),
            dict(type='Resize', img_scale=(512, 512), ratio_range=(0.5, 2.0)),
            dict(type='RandomCrop', crop_size=(512, 512), cat_max_ratio=0.75),
            dict(type='RandomFlip', prob=0.5),
            dict(type='PhotoMetricDistortion'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='Pad', size=(512, 512), pad_val=0, seg_pad_val=5),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img', 'gt_semantic_seg'])
        ]),
    val=dict(
        type='PotsdamDataset',
        data_root='/root/PycharmProjects/xmx/data/potsdam_all',
        img_dir='img_dir_IRRG/val',
        ann_dir='ann_dir_IRRG/val',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(512, 512),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]),
    test=dict(
        type='PotsdamDataset',
        data_root='/root/PycharmProjects/xmx/data/potsdam_all',
        img_dir='img_dir_IRRG/val',
        ann_dir='ann_dir_IRRG/val',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=None,
                img_ratios=[1.0],
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]))
log_config = dict(
    interval=50, hooks=[dict(type='TextLoggerHook', by_epoch=False)])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
cudnn_benchmark = True
optimizer = dict(
    type='AdamW',
    lr=6e-05,
    betas=(0.9, 0.999),
    weight_decay=0.01,
    paramwise_cfg=dict(
        custom_keys=dict(
            absolute_pos_embed=dict(decay_mult=0.0),
            relative_position_bias_table=dict(decay_mult=0.0),
            norm=dict(decay_mult=0.0))))
optimizer_config = dict()
lr_config = dict(
    policy='poly',
    warmup='linear',
    warmup_iters=1500,
    warmup_ratio=1e-06,
    power=1.0,
    min_lr=0.0,
    by_epoch=False)
runner = dict(type='IterBasedRunner', max_iters=80000)
checkpoint_config = dict(by_epoch=False, interval=2000)
evaluation = dict(interval=2000, metric=['mIoU', 'mFscore'], pre_eval=True)
find_unused_parameters = True
work_dir = './work_dirs/upernet_vitae_win_window7_512x512_80k_potsdam_epoch100'
gpu_ids = range(0, 2)
auto_resume = False


```

mIoU还是会出现几乎为0的情况

**Comment by Lee-inso** [comment]
> @Lee-inso 这个问题比较复杂，需要慢慢盘：
> 
> 1. 你用的是哪种数据和标签？
>    数据有，RGB, IRRG和RGBIR三种
>    标签有'5_Labels_all.zip'和'5_Labels_all_noBoundary.zip'两种
> 2. mmseg用'5_Labels_all_noBoundary.zip'，它首先将黑边转换到0，其余颜色转换到1-6，然后`reduce_zero_label=True`等一通操作后将黑色转到255，然后通过`ignore_index=255`在算loss时不考虑黑色，然后mmseg直接评测其余六个类别（包含clutter）的精度
> 3. 我们用'5_Labels_all.zip'，其中无黑色边界，因此我们直接将类别转到0-5（5是clutter），保持`reduce_zero_label=False`即可，因为遥感论文中通常不考虑cluster，所以我们`num_classes=5` `ignore_index=5`
> 4. 如果你用'5_Labels_all_noBoundary.zip'，用mmseg转完以后，会有包含黑色和clutter在内七个类别，这样无法同时ignore两个类别
> 5. 如果你用'5_Labels_all.zip'，用mmseg转完以后，因为无黑色，类别变成1-6，所以需要`reduce_zero_label=True`，变成0-5，然后你需要同时在config和potsdam_ori.py中(如果你用的话)设置`ignore_index=5`
> 
> 所以我猜，虽然你用了potsdam_ori.py， 但是没有设置`ignore_index=5`，影响了精度的计算

```
然后你需要同时在config和potsdam_ori.py中(如果你用的话)设置`ignore_index=5`
```
非常感谢，解答很清晰，5说的没错，我因为觉得potsdam.py里的ignore_index跟config中的ignore重复，所以给注释掉了
![image](https://user-images.githubusercontent.com/59754763/225787605-efe8ddad-1896-4564-85ee-ceeb241156b5.png)

为什么此两处均需要设置呢？

**Comment by 1835969208** [comment]
> > @Lee-inso 这个问题比较复杂，需要慢慢盘：
> > 
> > 1. 你用的是哪种数据和标签？
> >    数据有，RGB, IRRG和RGBIR三种
> >    标签有'5_Labels_all.zip'和'5_Labels_all_noBoundary.zip'两种
> > 2. mmseg用'5_Labels_all_noBoundary.zip'，它首先将黑边转换到0，其余颜色转换到1-6，然后`reduce_zero_label=True`等一通操作后将黑色转到255，然后通过`ignore_index=255`在算loss时不考虑黑色，然后mmseg直接评测其余六个类别（包含clutter）的精度
> > 3. 我们用'5_Labels_all.zip'，其中无黑色边界，因此我们直接将类别转到0-5（5是clutter），保持`reduce_zero_label=False`即可，因为遥感论文中通常不考虑cluster，所以我们`num_classes=5` `ignore_index=5`
> > 4. 如果你用'5_Labels_all_noBoundary.zip'，用mmseg转完以后，会有包含黑色和clutter在内七个类别，这样无法同时ignore两个类别
> > 5. 如果你用'5_Labels_all.zip'，用mmseg转完以后，因为无黑色，类别变成1-6，所以需要`reduce_zero_label=True`，变成0-5，然后你需要同时在config和potsdam_ori.py中(如果你用的话)设置`ignore_index=5`
> > 
> > 所以我猜，虽然你用了potsdam_ori.py， 但是没有设置`ignore_index=5`，影响了精度的计算
> 
> ```
> 然后你需要同时在config和potsdam_ori.py中(如果你用的话)设置`ignore_index=5`
> ```
> 
> 非常感谢，解答很清晰，3Q

请问一下这个postsdam_ori.py和config文件夹的完整目录地址能给一下吗
